crates/solver/src/settlement_rater.rs

```diff
@@ -259,3 +277,151 @@ impl SettlementRating for SettlementRater {
         Ok(Rating::Ok(rated_settlement))
     }
 }
+
+pub struct ScoreCalculator {
+    score_cap: BigRational,
+}
+
+impl ScoreCalculator {
+    pub fn new(score_cap: BigRational) -> Self {
+        Self { score_cap }
+    }
+
+    pub fn compute_score(
+        &self,
+        objective_value: &BigRational,
+        success_probability: f64,
+    ) -> Result<Score> {
+        ensure!(
+            (0.0..=1.0).contains(&success_probability),
+            "success probability must be between 0 and 1."
+        );
+        let success_probability =
+            BigRational::from_float(success_probability).context("Invalid success probability.")?;
+        let cost_fail = BigRational::from_float(0.0).unwrap();
+        let optimal_score =
+            self.compute_optimal_bid(objective_value.clone(), success_probability, cost_fail)?;
+        ensure!(
+            optimal_score <= *objective_value,
+            "Optimal score higher than objective value"
+        );
+        let score = big_rational_to_u256(&optimal_score).context("Invalid score.")?;
+        Ok(Score::Protocol(score))
+    }
+
+    fn compute_optimal_bid(
+        &self,
+        objective: BigRational,
+        probability_success: BigRational,
+        cost_fail: BigRational,
+    ) -> Result<BigRational> {
+        tracing::trace!(
+            ?objective,
+            ?probability_success,
+            ?cost_fail,
+            "Computing optimal bid"
+        );
+        let probability_fail = BigRational::one() - probability_success.clone();
+        let payoff_obj_minus_cap = self.payoff(
+            objective.clone() - self.score_cap.clone(),
+            objective.clone(),
+            probability_success.clone(),
+            cost_fail.clone(),
+        );
+        tracing::trace!(?payoff_obj_minus_cap, "Payoff obj minus cap");
+        let payoff_cap = self.payoff(
+            self.score_cap.clone(),
+            objective.clone(),
+            probability_success.clone(),
+            cost_fail.clone(),
+        );
+        tracing::trace!(?payoff_cap, "Payoff cap");
+        let optimal_bid = if payoff_obj_minus_cap >= BigRational::zero()
+            && payoff_cap <= BigRational::zero()
+        {
+            Ok(probability_success * objective - probability_fail * cost_fail)
+        } else if payoff_obj_minus_cap >= BigRational::zero() && payoff_cap > BigRational::zero() {
+            ensure!(!probability_success.is_zero(), "success is zero");
```

Comment:
I think it's weird that it's possible to not get a score out of `compute_optimal_bid()` when `probability_success == 0`. Intuitively I would say that result in a score of 0 as that is the lowest value.
This was probably handled here to avoid division by zero but maybe that case should rather be handled at the top of `compute_optimal_bid()` to avoid that edge case altogether?